### ISSUE TITLE
Failed to connect to network in Windows 10 system

### ISSUE THREAD

**Issue by zhouxinghong** [LorenEteval/Furious#24]
Failed to connect to network in Windows 10 system

Hello! In Windows 10 system, when you turn on VPN mode, you always fail to connect to the network!
Furious: 0.2.10

**Comment by zhouxinghong** [comment]
```html
[2023-10-19 19:00:55,352] [Furious.Core.Core] [INFO] Tun2socks 2.5.1.2 started
[2023-10-19 19:00:57,446] [Furious.Utility.RoutingTable] [INFO] add rule 13.215.602.605->192.168.12.189 to routing table success
[2023-10-19 19:00:57,667] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed
[2023-10-19 19:00:57,681] [Furious.Core.Core] [ERROR] Tun2socks stopped unexpectedly with exitcode 1
```

**Comment by LorenEteval** [comment]
Check out [TUN mode wiki](https://github.com/LorenEteval/Furious/wiki/TUN-Mode).

**Comment by zhouxinghong** [comment]
[2023-10-19 20:55:19,250] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed

**Comment by zhouxinghong** [comment]
[2023-10-19 21:04:43,834] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed
[2023-10-19 21:04:43,934] [Furious.Utility.Proxy] [INFO] set proxy server 127.0.0.1:10809 success

**Comment by LorenEteval** [comment]
> [2023-10-19 20:55:19,250] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed
> 
> After installing wintun.dll in C:\Windows\System32. as you provided, I still cannot connect to the network through VPN mode.

Try restart the application. If that does not work, try restart your PC and try again.

**Comment by zhouxinghong** [comment]
OK, thank you very much for your help! I'll try restarting the PC later

**Comment by zhouxinghong** [comment]
The VPN mode is unstable. Yesterday I was able to connect to the Internet in VPN mode after restarting my PC. However, today I used VPN mode to connect to the Internet again and found that the connection failed again. . .
[2023-10-19 21:32:24,094] [Furious.Utility.RoutingTable] [INFO] set device 'Furious' gateway address '10.0.68.1' success
[2023-10-20 06:25:30,067] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed

**Comment by zhouxinghong** [comment]
[2023-10-20 08:26:02,958] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed

**Comment by zhouxinghong** [comment]
Re steps:
1. Restart your computer;

2. Turn on the VPN mode connection and successfully access google.com;
[Furious.Utility.RoutingTable] [INFO] set device 'Furious' gateway address '10.0.68.1' success

3. Switch service nodes, for example, from Germany to Singapore;

4. Failed to enable VPN mode, the following information is output:
  [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed

**Comment by zhouxinghong** [comment]
Since the VPN mode fails to connect to the network, many applications cannot connect to the network.

**Comment by zhouxinghong** [comment]
After a period of time, if you switch the service node, you will still be unable to successfully connect to the network in VPN mode. . .

**Comment by LorenEteval** [comment]
> After a period of time, if you switch the service node, you will still be unable to successfully connect to the network in VPN mode. . . [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed

Hi. If Furious cannot set `10.0.68.1` as its own TUN gateway address then yes the VPN mode cannot work properly. Please make sure that no other application is using `10.0.68.1` as gateway address.

Also I just uploaded a new [release file](https://github.com/LorenEteval/Furious/releases/tag/0.2.10) `Furious-0.2.11-windows-x86_64-prerelease0.zip` that contains improved error messages for routing table operations. Please try using it in VPN mode and see what possible error messages that Furious gives. Feel free to paste them out here.

**Comment by zhouxinghong** [comment]
OK! Thank you so much! 👍😜
It should be noted that the TUN mode of Clash for Windows runs very stably on my machine, and its implementation principle should be similar to Furious. You can take a look at it.

**Comment by zhouxinghong** [comment]
could not find it Furious-0.2.11-windows-x86_64-prerelease0.zip

**Comment by LorenEteval** [comment]
> could not find it Furious-0.2.11-windows-x86_64-prerelease0.zip

![image](https://github.com/LorenEteval/Furious/assets/139033850/8d77cda7-22f7-41ec-8f20-25fc3ccf98b9)

**Comment by zhouxinghong** [comment]
Feel sorry! I didn’t see it just now😂

**Comment by zhouxinghong** [comment]
Feel sorry! I didn’t see it just now😂

**Comment by zhouxinghong** [comment]
After a while, switch the service node again, and the following message will appear:
[2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
[2023-10-20 18:19:33,086] [Furious.Utility.Proxy] [INFO] set proxy server 127.0.0.1:10809 success

Failed to connect to the network in VPN mode!
Furious-0.2.11-windows-x86_64-prerelease0
Windows 10 19044

**Comment by zhouxinghong** [comment]
```html
] [INFO] add rule tjcu-l6ftl9.zzcloud.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
[2023-10-20 18:19:32,575] [Furious.Utility.RoutingTable] [INFO] add rule 123.158.xxx.xxx->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: 操作完成!. Stderr: 
[2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
```

**Comment by LorenEteval** [comment]
> After turning on VPN mode and connecting to the network for a period of time and then switching service nodes again, the following message will appear:
> 
> ```
> ] [INFO] add rule tjcu-l6ftl9.zzcloud.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
> [2023-10-20 18:19:32,575] [Furious.Utility.RoutingTable] [INFO] add rule 123.158.xxx.xxx->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: 操作完成!. Stderr: 
> [2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
> [2023-10-20 18:19:33,086] [Furious.Utility.Proxy] [INFO] set proxy server 127.0.0.1:10809 success
> ```
> 
> ![image](https://user-images.githubusercontent.com/54253524/276902806-b89e7c0b-3c0f-4a29-9cfd-31bd9b2e9ca7.png)
> 
> Failed to connect to the network when turning on VPN mode! environment: Furious-0.2.11-windows-x86_64-prerelease0 Windows 10 19044 x64

Thank you for your reply. Discovered:

* `[INFO] add rule tjcu-l6ftl9.zzcloud.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误`. This is a bug in Furious that caused by resolved to another CNAME record,  which Furious doesn't handle.
* `[2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: `. I cannot reproduce on my side. It is probably caused by the `Furious` TUN device hasn't brought up yet. I'll add some guards before actually setting the device gateway address, to ensure that `Furious` TUN device exists. Not sure if this solves the problem.

I'll add another prerelease tomorrow, if you are kind enough to test it. Thank you.

**Comment by zhouxinghong** [comment]
Second case:
When the computer wakes up from sleep, running Furious to enable VPN mode will also output the above error message.
The new error message is as follows:
![image](https://github.com/LorenEteval/Furious/assets/54253524/af2dcf34-cf90-4230-926b-278704a2ae94)

**Comment by zhouxinghong** [comment]
> > After turning on VPN mode and connecting to the network for a period of time and then switching service nodes again, the following message will appear:
> > ```
> > ] [INFO] add rule tjcu-l6ftl9.zzcloud.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
> > [2023-10-20 18:19:32,575] [Furious.Utility.RoutingTable] [INFO] add rule 123.158.xxx.xxx->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: 操作完成!. Stderr: 
> > [2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
> > [2023-10-20 18:19:33,086] [Furious.Utility.Proxy] [INFO] set proxy server 127.0.0.1:10809 success
> > ```
> > 
> > 
> > ![image](https://user-images.githubusercontent.com/54253524/276902806-b89e7c0b-3c0f-4a29-9cfd-31bd9b2e9ca7.png)
> > Failed to connect to the network when turning on VPN mode! environment: Furious-0.2.11-windows-x86_64-prerelease0 Windows 10 19044 x64
> 
> Thank you for your reply. Discovered:
> 
> * `[INFO] add rule tjcu-l6ftl9.zzcloud.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误`. This is a bug in Furious that caused by resolved to another CNAME record,  which Furious doesn't handle.
> * `[2023-10-20 18:19:32,982] [Furious.Utility.RoutingTable] [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: `. I cannot reproduce on my side. It is probably caused by the `Furious` TUN device hasn't brought up yet. I'll add some guards before actually setting the device gateway address, to ensure that `Furious` TUN device exists. Not sure if this solves the problem.
> 
> I'll add another prerelease tomorrow focusing on these two issues, if you are kind enough to test it. Thank you.

OK! no problem.

**Comment by LorenEteval** [comment]
> OK! no problem.
> 
> ### Second case:
> When the computer wakes up from sleep, running Furious to enable VPN mode will also output the following error message: ![image](https://user-images.githubusercontent.com/54253524/276937320-af2dcf34-cf90-4230-926b-278704a2ae94.png)
> 
> ```
> [INFO] add rule hkcu.relay-idc.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
>  [INFO] add rule apyd1.xn--nmq78hhd72yv8gyw4g.com.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
>  [INFO] add rule 36.133.xxx.xxx->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: 操作完成!. Stderr: 
>  [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
>  [INFO] set proxy server 127.0.0.1:10809 success
>  [INFO] action is 'Show Log...'. Checked is False
> ```
> 
> Once again, VPN mode failed to connect to the network!

Added `Furious-0.2.11-windows-x86_64-prerelease1.zip` in [release page](https://github.com/LorenEteval/Furious/releases/tag/0.2.10). Please check if the problem still exists.

**Comment by zhouxinghong** [comment]
> > OK! no problem.
> > ### Second case:
> > When the computer wakes up from sleep, running Furious to enable VPN mode will also output the following error message: ![image](https://user-images.githubusercontent.com/54253524/276937320-af2dcf34-cf90-4230-926b-278704a2ae94.png)
> > ```
> > [INFO] add rule hkcu.relay-idc.top.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
> >  [INFO] add rule apyd1.xn--nmq78hhd72yv8gyw4g.com.->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: . Stderr: 路由添加失败: 参数错误。
> >  [INFO] add rule 36.133.xxx.xxx->192.168.xx.xxx to routing table success. Returncode: 0. Stdout: 操作完成!. Stderr: 
> >  [ERROR] set device 'Furious' gateway address '10.0.68.1' failed. Returncode: 1. Stdout: 找不到元素。. Stderr: 
> >  [INFO] set proxy server 127.0.0.1:10809 success
> >  [INFO] action is 'Show Log...'. Checked is False
> > ```
> > 
> > 
> > Once again, VPN mode failed to connect to the network!
> 
> Added `Furious-0.2.11-windows-x86_64-prerelease1.zip` in [release page](https://github.com/LorenEteval/Furious/releases/tag/0.2.10). Please check if the problem still exists.

OK! no problem.

**Comment by zhouxinghong** [comment]
Switching from one node to the following node outputs the following information:
```
vmess://eyJ2IjoiMiIsInBzIjoiJUYwJTlGJTg3JUFGJUYwJTlGJTg3JUI1JTIwMyVDMyU5NyUyMCVFNiU5NyVBNSVFNiU5QyVBQyUyMElQTEMiLCJhZGQiOiJqcGlwbGMucmVsYXktaWRjLnRvcCIsInBvcnQiOjI2MDgwLCJpZCI6IjUxMDU0ZWMxLTg2NDUtNDYxMy04MTY5LWZkYTU4MjdjNzVmYiIsImFpZCI6MCwic2N5IjoiYXV0byIsIm5ldCI6IndzIiwidGxzIjoibm9uZSIsInBhdGgiOiIlMkZnaXRlZSIsImhvc3QiOiJqcGlwbGMucmVsYXktaWRjLnRvcCJ9
```

```
[2023-10-21 16:40:07,701] [Furious.Core.Core] [INFO] Xray-core 1.8.4.1 started
[2023-10-21 16:40:11,070] [Furious.Utility.RoutingTable] [INFO] get default gateway address success. ['192.xxx.xx.xxx']
[2023-10-21 16:40:11,655] [Furious.Utility.Utility] [INFO] DNS resolution uses proxy server 127.0.0.1:10809
[2023-10-21 16:40:11,658] [Furious.Utility.Utility] [INFO] DNS resolution in progress. Wait
[2023-10-21 16:40:16,234] [Furious.Utility.Utility] [ERROR] DNS resolution for 'jpiplc.relay-idc.top' failed. Connection closed
[2023-10-21 16:40:16,333] [Furious.Utility.Proxy] [INFO] turn off proxy success
[2023-10-21 16:40:16,335] [Furious.Core.Core] [INFO] Xray-core terminated with exitcode -15
```
I'm not sure if this is Furious's wrong behavior. You can help verify whether the node is really unable to connect to the network. . .

**Comment by LorenEteval** [comment]
Solved and will come in official version 0.2.11.
